src/files/normalise-input.js

```diff
@@ -0,0 +1,233 @@
+'use strict'
+
+const errCode = require('err-code')
+const { Buffer } = require('buffer')
+
+/*
+ * Transform one of:
+ *
+ * ```
+ * Buffer|ArrayBuffer|TypedArray
+ * Blob|File
+ * { path, content: Blob }
+ * { path, content: String }
+ * { path, content: Iterable<Number> }
+ * { path, content: Iterable<Buffer> }
+ * { path, content: Iterable<Iterable<Number>> }
+ * { path, content: AsyncIterable<Iterable<Number>> }
+ * String
+ * Iterable<Number>
+ * Iterable<Buffer>
+ * Iterable<Blob>
+ * Iterable<{ path, content: Buffer }>
+ * Iterable<{ path, content: Blob }>
+ * Iterable<{ path, content: Iterable<Number> }>
+ * Iterable<{ path, content: AsyncIterable<Buffer> }>
+ * AsyncIterable<Buffer>
+ * AsyncIterable<{ path, content: Buffer }>
+ * AsyncIterable<{ path, content: Blob }>
+ * AsyncIterable<{ path, content: Iterable<Buffer> }>
+ * AsyncIterable<{ path, content: AsyncIterable<Buffer> }>
+ * ```
```

Comment:
I was hoping to not include it here. My thinking was that all `ipfs.add*` methods could route to `ipfs._addAsyncIterator`, and it would be down to those methods to convert their arguments to something this method accepts.  E.g. `ipfs.addPullStream` would do the `pull-stream` source to async iterator conversion, `ipfs.addFromFs` would convert glob-matched paths to async iterators of file objects with paths & content readable streams, etc.

I'm working on integrating that with `ipfs` and `ipfs-http-client` so we'll see how realistic that is..

Reply:
That sounds nice, but you can call `ipfs.add({ content: PullStreamSource })` so I think it's necessary, unless we want to duplicate a lot of the normalise work in both places.